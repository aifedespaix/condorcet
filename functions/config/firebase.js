const admin = require("firebase-admin");

const serviceAccount = require("./serviceAccountKey.json");

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: "https://firedev-11a28.firebaseio.com"
});

const firestore = admin.firestore();
const settings = {
  timestampsInSnapshots: true
};
firestore.settings(settings);

exports.db = firestore;
exports.auth = admin.auth();
