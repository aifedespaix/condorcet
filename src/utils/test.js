import _get from "lodash/get";
import frLocal from "../locales/fr.json";

export const MOCK_POLL_ID = "MOCK_POLL_ID";

export const MOCK_USER_ID = "MOCK_USER_ID";

export const MOCK_TOKEN = "MOCK_TOKEN";

export const MOCK_TIMESTAMP = { seconds: 1559325107, nanoseconds: 154000000 };

export const MOCK_ANSWERS = [
  { slug: "luffy", value: "Luffy" },
  { slug: "zorro", value: "Zorro" },
  { slug: "sanji", value: "Sanji" }
];

export const MOCK_POLL = {
  isActive: true,
  isOwner: false,
  answers: MOCK_ANSWERS,
  description: "Qui repellendus est aliquam vel qui.",
  isPublic: true,
  owner: MOCK_USER_ID,
  question: "Qui est le meilleur ?",
  timestamp: MOCK_TIMESTAMP,
  token: MOCK_TOKEN,
  users: [MOCK_USER_ID]
};

export const getTranslation = path => _get(frLocal, path);
