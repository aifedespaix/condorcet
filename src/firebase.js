import firebase from "@firebase/app";
import "@firebase/firestore";
import "@firebase/auth";

const config = {
  apiKey: "AIzaSyBl7KGWaRY0B-BQ7WhUwSnmFHj3f3OYF0U",
  authDomain: "condorcet-4167c.firebaseapp.com",
  databaseURL: "https://condorcet-4167c.firebaseio.com",
  projectId: "condorcet-4167c",
  storageBucket: "condorcet-4167c.appspot.com",
  messagingSenderId: "400396103060"
};

firebase.initializeApp(config);

// Db
export const db = firebase.firestore();
db.settings({ timestampsInSnapshots: true });

// Auth
export const auth = firebase.auth();

export default firebase;
