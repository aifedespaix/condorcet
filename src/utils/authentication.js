import firebase, { auth } from "../firebase";

const provider = new firebase.auth.GoogleAuthProvider();

export const signIn = cb => {
  auth.useDeviceLanguage();
  auth
    .signInWithPopup(provider)
    .then(result => {
      auth.setPersistence(firebase.auth.Auth.Persistence.LOCAL);
      cb(result.user);
    })
    .catch(function(error) {
      console.error(error);
    });
};

export const signOut = cb => {
  auth
    .signOut()
    .then(() => {
      cb(null);
    })
    .catch(error => {
      cb({ status: "error", error });
    });
};
