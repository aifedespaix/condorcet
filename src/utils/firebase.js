import { db } from "../firebase";

class Firebase {
  constructor(collection, doc = null, subcollection = null) {
    this.collection =
      doc && subcollection
        ? db
            .collection(collection)
            .doc(doc)
            .collection(subcollection)
        : db.collection(collection);
  }

  /**
   * Prend une list snapshot de document et retourne les données au format JSON
   * @param {*} docs
   */
  static convertDocumentsSnapshotToJson(docs) {
    return docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
  }

  /**
   * Prend une list snapshot de document et retourne les données au format JSON
   * @param {*} doc
   */
  static convertDocumentSnapshotToJson(doc) {
    return { id: doc.id, ...doc.data() };
  }

  /**
   * Retourne l'instance
   */
  getInstance() {
    return this.collection;
  }

  /**
   * Ajout un document a la collection
   * @param {*} data
   */
  async create(data) {
    const newDocument = await this.collection.add(data);
    const newDocumentSnapshot = await newDocument.get();

    return Firebase.convertDocumentSnapshotToJson(newDocumentSnapshot);
  }

  /**
   * Liste les documents d'une collection
   * @param {*} whereClause Paramètre optionnel ajoutant une clause where sur la requête
   */
  async list(whereClause = { field: null, operator: null, value: null }) {
    const { field, operator, value } = whereClause;

    if (field && operator && value) {
      return Firebase.convertDocumentsSnapshotToJson(
        (await this.collection.where(field, operator, value).get()).docs
      );
    }

    return Firebase.convertDocumentsSnapshotToJson(
      (await this.collection.get()).docs
    );
  }

  /**
   * Retourne un document basé sur son identifiant
   * @param {*} id
   */
  async get(id) {
    const document = await this.collection.doc(id).get();

    return Firebase.convertDocumentSnapshotToJson(document);
  }

  // TODO: return OK response
  update(id, data) {
    this.collection.doc(id).update(data);
  }

  /**
   * Supprime un document
   * @param {*} id
   */
  async delete(id) {
    try {
      await this.collection.doc(id).delete();

      return id;
    } catch (error) {
      return { error: true, detail: error };
    }
  }
}

export default Firebase;
