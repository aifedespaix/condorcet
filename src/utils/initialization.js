import { auth } from "../firebase";
import { INITIALIZE } from "../types";

/**
 * isAppInitialized: bool
 * cb(user)
 */
export default (store, cb) => {
  const isAppInitialized = store.state.app.initialized;

  if (isAppInitialized === true) {
    cb({ state: "already_initialized" });
    return;
  }

  const timeoutId = setTimeout(() => {
    store.commit(`app/${INITIALIZE}`);
    cb({ state: "no_user" });
  }, 1500);

  auth.onAuthStateChanged(user => {
    if (!user) return;

    if (timeoutId) {
      clearTimeout(timeoutId);
    }

    store.commit(`app/${INITIALIZE}`);
    store.dispatch("user/setUser", user);

    cb({ state: "just_logged" });
  });
};
