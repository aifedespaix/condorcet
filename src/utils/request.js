import firebase, { db } from "../firebase";
import { VoteModel } from "../models/Vote";
import { PollModel } from "../models/Poll";
import { validate } from "indicative";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  const validatedPoll = await validate(pollWithTimestap, PollModel);

  return db.collection("polls").add(validatedPoll);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  const validatedVote = await validate(voteWithTimestap, VoteModel);

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(validatedVote);
};
