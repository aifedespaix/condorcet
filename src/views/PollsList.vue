<template>
  <div>
    <v-layout row wrap>
      <v-flex xs12 sm6 v-for="poll in polls" :key="poll.id">
        <v-card>
          <v-card-title primary-title>
            <div>
              <h3 class="headline mb-0">{{ poll.name }}</h3>
              <div>
                {{ poll.description }}
              </div>
            </div>
          </v-card-title>
          <v-card-actions>
            <v-btn flat color="orange" @click.prevent="copyMagicLink(poll.id)"
              >Partager</v-btn
            >
            <v-btn
              v-if="poll.isOwner"
              flat
              color="orange"
              @click.prevent="removePoll(poll.id, poll.name)"
              >Supprimer</v-btn
            >
            <v-btn
              v-if="poll.isOwner"
              flat
              color="orange"
              @click.prevent="togglePoll(poll.id, poll.name)"
              >{{ poll.isActive ? "Fermer" : "Ré-ouvir" }}</v-btn
            >
            <!-- <v-btn  flat color="orange" @click.prevent="togglePoll(poll.id, poll.name)">{{ poll.isActive ? "Fermer" : "Ré-ouvir" }}</v-btn> -->

            <router-link
              class="v-btn v-btn--flat theme--light orange--text"
              tag="a"
              :to="{
                name: poll.isActive ? 'poll_show' : 'poll_result',
                params: { id: poll.id }
              }"
              >{{ poll.isActive ? "Voter" : "Résultats" }}</router-link
            >
          </v-card-actions>
        </v-card>
      </v-flex>
    </v-layout>
  </div>
</template>

<script>
import { mapActions, mapState, mapGetters } from "vuex";
import { db } from "@/firebase";

export default {
  name: "PollsList",
  mounted() {
    this.fetchPolls(this.userId);
  },
  computed: {
    ...mapGetters("user", ["userId"]),
    ...mapState({
      polls: state => state.poll.all
    })
  },
  methods: {
    ...mapActions("app", ["addNotification"]),
    ...mapActions("poll", ["fetchPolls"]),
    async togglePoll(id) {
      try {
        const document = db.collection("polls").doc(id);
        const currentPollState = (await document.get()).data();

        const newPollState = {
          ...currentPollState,
          isActive: !currentPollState.isActive
        };

        document.set(newPollState);
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème`,
          status: "error"
        });
      }
    },
    removePoll(id, name) {
      try {
        db.collection("polls")
          .doc(id)
          .delete();
        this.addNotification({
          text: `La salle ${name} a bien été supprimée`,
          status: "info"
        });
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème lors de la suppression de ${name}`,
          status: "error"
        });
      }
    },
    async copyMagicLink(id) {
      const poll = this.polls.find(poll => poll.id === id);
      const link = `${window.location.origin}/polls/${poll.id}/join?token=${
        poll.token
      }`;

      try {
        await this.$copyText(link);
        this.addNotification({
          text: `Lien copié`,
          status: "info"
        });
      } catch (error) {
        this.addNotification({
          text: `Lien copié`,
          status: "error"
        });
      }
    }
  }
};
</script>

<style></style>
