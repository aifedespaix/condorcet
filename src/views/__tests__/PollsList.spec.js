import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";

import { db } from "../../firebase";
import { MOCK_USER_ID, MOCK_POLL_ID } from "../../utils/test";
import PollsList from "../PollsList.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock("../../firebase");

describe("PollsList.vue", () => {
  let wrapper;
  let appActions;
  let pollActions;
  let pollGetters;

  beforeEach(() => {
    jest.clearAllMocks();

    pollActions = {
      listenPolls: jest.fn(),
      leavePoll: jest.fn(),
      removePoll: jest.fn()
    };

    appActions = {
      addNotification: jest.fn()
    };

    pollGetters = {
      userPolls: jest.fn(() => [])
    };

    const store = new Vuex.Store({
      modules: {
        app: {
          namespaced: true,
          actions: appActions
        },
        user: {
          namespaced: true,
          getters: {
            userId: () => MOCK_USER_ID
          }
        },
        poll: {
          namespaced: true,
          actions: pollActions,
          getters: pollGetters
        }
      }
    });

    wrapper = shallowMount(PollsList, {
      store,
      stubs: ["router-link"],
      localVue
    });
  });

  it("should render with message saying there is no polls", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with one poll", () => {
    pollGetters.userPolls;
    expect(wrapper).toMatchSnapshot();
  });

  it("should leave poll", () => {
    window.confirm = jest.fn(() => true);
    wrapper.vm.leavePoll(MOCK_POLL_ID);
    expect(db.collection).toBeCalledWith("polls");
    expect(db.collection("polls").doc).toBeCalledWith(MOCK_POLL_ID);
  });
  it("should not leave poll", () => {
    window.confirm = jest.fn(() => false);
    wrapper.vm.leavePoll(MOCK_POLL_ID);
    expect(db.collection).not.toBeCalled();
    expect(db.collection("polls").doc).not.toBeCalled();
  });
});
