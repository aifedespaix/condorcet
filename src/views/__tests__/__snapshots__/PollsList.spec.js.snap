// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollsList.vue should render with message saying there is no polls 1`] = `
<loader-stub condition="true">
  <div class="flex">
    <div>
      Tu n'a pas encore de scrutin. Tu peux en créé un en cliquant sur
      <router-link-stub to="polls/new" class="text-teal-500 hover:underline">ce lien</router-link-stub>.
    </div>
  </div>
</loader-stub>
`;

exports[`PollsList.vue should render with one poll 1`] = `
<loader-stub condition="true">
  <div class="flex">
    <div>
      Tu n'a pas encore de scrutin. Tu peux en créé un en cliquant sur
      <router-link-stub to="polls/new" class="text-teal-500 hover:underline">ce lien</router-link-stub>.
    </div>
  </div>
</loader-stub>
`;
